WordPress/Classes/Services/MediaService.m

```diff
@@ -56,6 +60,141 @@ - (id)initWithManagedObjectContext:(NSManagedObjectContext *)context
+                               }
+                           }];
+}
+
+- (void)mergeMedia:(NSArray *)media forBlog:(Blog *)blog completionHandler:(void (^)(void))completion
+{
+    NSMutableArray *mediaToKeep = [NSMutableArray array];
+    for (RemoteMedia *remote in media) {
+        Media *local = [self findMediaWithID:remote.mediaID inBlog:blog];
+        if (!local) {
+            local = [self newMediaForBlog:blog];
+            local.remoteStatus = MediaRemoteStatusSync;
+        }
+        [self updateMedia:local withRemoteMedia:remote];
+        [mediaToKeep addObject:local];
+    }
+    
+    NSSet *existingMedia = blog.media;
+    NSMutableOrderedSet *mediaToThumbnail = [NSMutableOrderedSet orderedSet];
+    if (existingMedia.count > 0) {
+        for (Media *existing in existingMedia) {
+            if (![mediaToKeep containsObject:existing] && existing.remoteURL != nil) {
+                [self.managedObjectContext deleteObject:existing];
+            } else if (existing.remoteURL.length && !existing.thumbnail && existing.mediaType == MediaTypeImage) {
+                [mediaToThumbnail addObject:existing];
+            }
+        }
+    }
+    
+    [[ContextManager sharedInstance] saveDerivedContext:self.managedObjectContext];
+    
+    [mediaToThumbnail sortUsingComparator:^NSComparisonResult(Media *left, Media *right) {
+        return [right.creationDate compare:left.creationDate];
+    }];
+    [self getThumbnailsForMedia:mediaToThumbnail];
+    
+    if (completion) {
+        dispatch_async(dispatch_get_main_queue(), completion);
+    }
+}
+
+- (void)getThumbnailsForMedia:(NSMutableOrderedSet *)media
+{
+    Media *item = nil;
+    do {
+        item = media.firstObject;
+        [media removeObject:item];
+        // might have updated it to show in browser since last sync
+        [self.managedObjectContext refreshObject:item mergeChanges:YES];
+        if (!item.remoteURL.length || item.thumbnail || item.mediaType != MediaTypeImage) {
+            item = nil;
+            continue;
+        }
+    } while (!item && [media count]);
+    if (!item) {
+        return;
+    }
+
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0),^{
+        [self getThumbnailForMedia:item
+                           success:^(UIImage *image) {
+                               [self performSelector:@selector(getThumbnailsForMedia:) withObject:media afterDelay:0.1];
+                           }
+                           failure:^(NSError *error) {
+                               DDLogError(@"Failed getting thumbnail image: %@", error);
+                           }];
+    });
+}
+
+- (void)getThumbnailForMedia:(Media *)media
+                     success:(void (^)(UIImage *image))success
+                     failure:(void (^)(NSError *error))failure
+{
+    NSInteger thumbnailSize = round([WPBlogMediaCollectionViewController thumbnailWidthFor:[[UIScreen mainScreen] bounds].size.width]);
```

Comment:
Having a Service Object depending of a ViewController is not very good. Could we we implement this in another way?